design/feature-flags.md

```diff
@@ -0,0 +1,98 @@
+# "Future" Flags
+
+Sometimes we want to introduce new breaking behavior because we believe this is
+the correct default behavior for the CDK. The problem, of course, is that
+breaking changes are only allowed in major versions and those are rare.
+
+This document describes a proposal for a pattern/mechanism called "future
+flags". It will allow us to introduce breaking behavior which is disabled by
+default (so existing projects will not be affected) but enabled automatically
+for new projects created through `cdk init`.
+
+> The name "**future flags**" is a wordplay on "feature flags" (dah!) to
+> represent features that are "from the future" and indicate that these specific
+> feature flags are only disabled by default only until the next release, after
+> which they will either be removed _or_ turned into regular feature flags that
+> people can select if they so desire.
+>
+> Having said all that, I am okay if we feel this is unnecessary and we can call
+> these just "feature flags" and not have a semantic distinction between the
+> ones that are "temporary" and permanent.
+
+## Approach
+
+The basic idea is that new breaking behavior will always be disabled by default
+and only enabled when a certain CDK context parameter is set. If not enabled,
+the system will continue to behave exactly like it used to without breaking any
+existing projects.
+
+When we release a new major version of the AWS CDK, we will flip this behavior
+or completely remove the legacy behavior.
+
+In order for new projects to pick up this new behavior automatically, we will
+modify `cdk init` to inject the set of "future flags" into the generated
+`cdk.json` file. This means that the new project will have the latest behavior,
+but projects that were created prior to the introduction of this feature will
+have the same legacy behavior based on the set of capabilities that were
+available at the time of the project's creation. This list will be cleaned up
+every time we release a major version of course.
+
+Using fine-grained flags will allow users of old projects to pick up specific
+new "future" behaviors by manually adding the specific keys to their `cdk.json`
+file, without risking breakage in other unexpected areas.
+
+## Alternative Considered
+ 
+We considered an alternative of "bundling" new capabilities under a single flag
+that specifies the CDK version which created the project, but this means that
+users won't have the ability to pick and choose which future capabilities they
```

Comment:
Not exactly true, the 2 alternatives could be combined, as I described on the other PR.

Reply:
I know, but it feels like over-designing this without having sufficient evidence that it's required. This is not a one-way-door. We can always add "feature groups" later which will enable a whole bunch of feature flags with a single switch (and it could be a version number if we decide it's useful).

Let's start with the basic granular and simple mechanism and then evolve it as need arises, ok?

Reply:
Added a new under a new section for "future development"